homeassistant/components/xiaomi_aqara.py

```diff
@@ -189,23 +195,33 @@ def remove_device_service(call):
 class XiaomiDevice(Entity):
     """Representation a base Xiaomi device."""
 
-    def __init__(self, device, name, xiaomi_hub):
+    def __init__(self, device, name, hass, xiaomi_hub):
         """Initialize the xiaomi device."""
         self._state = None
+        self._lock = threading.Lock()
+        self._hass = hass
+        self._is_available = True
         self._sid = device['sid']
         self._name = '{}_{}'.format(name, self._sid)
         self._write_to_hub = xiaomi_hub.write_to_hub
         self._get_from_hub = xiaomi_hub.get_from_hub
         self._device_state_attributes = {}
+        self._remove_unavailability_tracker = None
+        self._track_unavailable()
         xiaomi_hub.callbacks[self._sid].append(self.push_data)
-        self.parse_data(device['data'])
+        self.parse_data(device['data'], device['raw_data'])
         self.parse_voltage(device['data'])
 
     @property
     def name(self):
         """Return the name of the device."""
         return self._name
 
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._is_available
```

Comment:
Can you not just do:
`return utcnow() < self._last_updated + TIME_TILL_UNAVAILABLE`
And set the _last_updated each time you hear from the device?

Reply:
Can you explain your point with more details? 
If you think this solution will give opportunity to remove `_track_unavailable()`, then you forgot `available()` is called only when we ask entity to update the state (`self.schedule_update_ha_state()`), and we ask for state update, only when device is alive and sends data. We need some observer that will do time tracking.
If you want to just get rid of `_is_available` variable, then from my point of view your solution is worse, because there will be 2 places which decide unavailability state - `_track_unavailable()`, which will trigger `_set_unavailable() -> schedule_update_ha_state()` and `available()` which will check condition for the 2nd time. It's better to have 1 place with logic (`_track_unavailable()`) and 2nd one (`available()`) which will just return the result.

Reply:
@Danielhiversen ^^

Reply:
I had hope to avoid using _track_unavailable(), but I see your point.

I am not sure this is a good solution, so I will let someone else review this pull request. Maybe they can come up with a better solution.

Reply:
@Danielhiversen 
> I will let someone else review this pull request
How will you do this?

Reply:
We need to track it ourselves for each device @Danielhiversen because the devices are `should_poll` False